azkaban-common/src/main/java/azkaban/executor/ExecutableRamp.java

```diff
@@ -293,143 +186,244 @@ public Metadata clone() {
-      }
+  synchronized public void rampDown() {
+    lock.lock();
+    try {
+      int currentStage = this.state.rampStage;
+      this.state.rampStage = currentStage - 1;
+      this.state.lastRampDownTime = System.currentTimeMillis();
+      this.state.isSynchronized = false;
+    } finally {
+      lock.unlock();
     }
+  }
+
+  synchronized public void cacheResult(Action action) {
+    lock.lock();
+    try {
+      this.state.cachedNumOfTrail++;
+      switch (action) {
+        case SUCCEEDED:
+          this.state.cachedNumOfSuccess++;
+          break;
+        case FAILED:
+          this.state.cachedNumOfFailure++;
+          break;
+        default:
+          this.state.cachedNumOfIgnored++;
+          break;
+      }
+      this.state.lastUpdatedTime = System.currentTimeMillis();
+
+      // verify the failure threshold
+      int trails = this.state.numOfTrail + this.state.cachedNumOfTrail;
+      int fails = this.state.numOfFailure + this.state.cachedNumOfFailure;
+      int failure =
+          this.metadata.isPercentageScaleForMaxFailure ? (trails == 0) ? 100 : (int) ((fails * 100.0) / (trails * 1.0))
+              : fails;
+
+      LOGGER.info(
+          "[Ramp Cached Result] (id = {}, action: {}, {} failure: {}, numOfTrail ({}, {}), numOfSuccess: ({}, {}), numOfFailure: ({}, {}), numOfIgnore: ({}, {}))",
+          this.id, action.name(), this.metadata.isPercentageScaleForMaxFailure ? "Percentage" : " ", failure, this.state.numOfTrail, this.state.cachedNumOfTrail, this.state.numOfSuccess, this.state.cachedNumOfSuccess,
+          this.state.numOfFailure, this.state.cachedNumOfFailure, this.state.numOfIgnored, this.state.cachedNumOfIgnored);
+
+      if (this.metadata.maxFailureToRampDown != 0) {
+        if (failure > this.metadata.maxFailureToRampDown) {
+          if (this.state.rampStage > 0) {
+            if (TimeUtils.timeEscapedOver(this.state.lastRampDownTime, ONE_DAY)) {
+              int currentStage = this.state.rampStage;
+              this.rampDown();
+              LOGGER.warn("[RAMP DOWN] (rampId = {}, failure = {}, threshold = {}, from stage {} to stage {}.)", this.getId(), failure, this.metadata.maxFailureToRampDown, currentStage, this.state.rampStage);
+            }
+          }
+        }
+      }
 
-    this.getState().markChanged();
+      if (this.metadata.maxFailureToPause != 0) {
+        if (failure > this.metadata.maxFailureToPause) {
+          this.state.isPaused = true;
+          LOGGER.info("[RAMP STOP] (rampId = {}, failure = {}, threshold = {}, timestamp = {})", this.getId(), failure,
+              this.metadata.maxFailureToPause, System.currentTimeMillis());
+        }
+      }
+
+      this.state.isSynchronized = false;
+    } finally {
+      lock.unlock();
+    }
   }
 
   synchronized public void cacheSaved() {
-    this.getState().markDataSaved();
+    lock.lock();
+    try {
+      int ttlTrail = this.state.numOfTrail + this.state.cachedNumOfTrail;
```

Comment:
We should protect that. The ramp status loading from db is independent to the ramp up/down. It could be possible one running flow trigged the ramp up/down and in the meantime the polling service is trying to load the status. Let's keep the lock to make sure there is no conflict.